### PR TITLE
Threadpool

### PR DESCRIPTION
**Pull request by moneromooo-monero**

### CODE REVIEW

File: src/common/threadpool.cpp [hunk 1]
```diff
@@ -73,12 +72,12 @@ void threadpool::submit(waiter *obj, std::function<void()> f) {
   } else {
     if (obj)
       obj->inc();
-    queue.push_back(e);
+    queue.push_back({obj, f});
```
Comment (hyc) [inline src/common/threadpool.cpp]: Was this change necessary? The original code is allocating early outside the lock to minimize time in the lock.

File: src/common/threadpool.cpp [hunk 2]
```diff
@@ -119,13 +118,12 @@ void threadpool::waiter::dec() {
...
   boost::unique_lock<boost::mutex> lock(mutex);
   while (running) {
-    entry e;
     while(queue.empty() && running)
       has_work.wait(lock);
     if (!running) break;
 
     active++;
-    e = queue.front();
+    entry e = queue.front();
```
Comment (hyc) [inline src/common/threadpool.cpp]: not seeing the reason for this change either.

File: src/common/threadpool.h [hunk 3]
```diff
@@ -46,6 +46,9 @@ class threadpool
     static threadpool instance;
     return instance;
   }
+  static threadpool *getNew(unsigned max_threads = 0) {
```
Comment (hyc) [inline src/common/threadpool.h]: This function shouldn't exist. The point of the threadpool is to be a singleton that manages all threads, instead of having threads created willy-nilly everywhere, like the old code did.

File: src/common/threadpool.cpp (hunk 1 above)
Comment (moneromooo-monero) [reply src/common/threadpool.cpp]: No, this was to avoid a temp object, I will revert.

File: src/common/threadpool.cpp (hunk 2 above)
Comment (moneromooo-monero) [reply src/common/threadpool.cpp]: same

File: src/common/threadpool.h (hunk 3 above)
Comment (moneromooo-monero) [reply src/common/threadpool.h]: This is used just for the tests, which use different number of threads.
